frameworklauncher/src/main/java/com/microsoft/frameworklauncher/common/model/ResourceDescriptor.java

```diff
@@ -143,35 +163,113 @@ public static ResourceDescriptor fromResource(Resource res) throws Exception {
+    }
+    LOGGER.logDebug("Get ResourceDescriptor: " + rd + " from hadoop resource: " + res);
     return rd;
   }
 
   public Resource toResource() throws Exception {
     Resource res = Resource.newInstance(memoryMB, cpuNumber);
+    Class<?> clazz = res.getClass();
 
     if (gpuNumber > 0) {
       try {
-        Class<?> clazz = res.getClass();
         Method setGpuNumber = clazz.getMethod("setGPUs", int.class);
         Method setGpuAttribute = clazz.getMethod("setGPUAttribute", long.class);
 
         setGpuNumber.invoke(res, gpuNumber);
         setGpuAttribute.invoke(res, gpuAttribute);
+
+      } catch (NoSuchMethodException e) {
+        LOGGER.logWarning(e, "Ignore: Failed to set GPU information, YARN library doesn't support");
+      } catch (IllegalAccessException e) {
+        LOGGER.logError(e, "Ignore: Failed to set GPU information, illegal access function");
+      }
+    }
+
+    if (portRanges != null && portRanges.size() > 0) {
+      try {
+
+        Class hadoopValueRangesClass = Class.forName("org.apache.hadoop.yarn.api.records.ValueRanges");
+        Class hadoopValueRangeClass = Class.forName("org.apache.hadoop.yarn.api.records.ValueRange");
+
+        Object obj = hadoopValueRangesClass.getMethod("newInstance").invoke(null);
+
+        List<Object> hadoopValueRangeList = new ArrayList<Object>();
+
+        for (Range range : portRanges) {
+          Object valueRangeObj = hadoopValueRangeClass.getMethod("newInstance", int.class, int.class).invoke(null, range.getBegin(), range.getEnd());
+          hadoopValueRangeList.add(valueRangeObj);
+        }
+
+        Method setRangesList = hadoopValueRangesClass.getMethod("setRangesList", List.class);
+        setRangesList.invoke(obj, hadoopValueRangeList);
+
+        Method setPort = clazz.getMethod("setPorts", hadoopValueRangesClass);
+        setPort.invoke(res, obj);
+
       } catch (NoSuchMethodException e) {
-        LOGGER.logWarning(e, "Ignore: Fail to set GPU information, YARN library doesn't support:");
+        LOGGER.logDebug(e, "Ignore: Failed to get Ports information, YARN library doesn't support Port");
       } catch (IllegalAccessException e) {
-        LOGGER.logError(e, "Ignore: Fail to set GPU information, illegal access function");
+        LOGGER.logError(e, "Ignore: Failed to get Ports information, illegal access function");
+      } catch (ClassNotFoundException e) {
+        LOGGER.logDebug(e, "Ignore: Failed to get the class Name");
+      } catch (Exception e) {
+        LOGGER.logDebug(e, "Ignore: Unknown exception");
       }
     }
+    LOGGER.logDebug("Put ResourceDescriptor " + this.toString() + " to hadoop resource: " + res);
     return res;
   }
 
   @Override
   public String toString() {
+    String portString = "";
+    if (portRanges != null) {
+      for (Range range : portRanges) {
+        portString = portString + "[" + range.getBegin() + "-" + range.getEnd() + "],";
+      }
+    }
```

Comment:
Use CommonExts.toString

Reply:
Better to use convertRangeToString?

---
In reply to: [167155269](https://github.com/Microsoft/pai/pull/318#discussion_r167155269) [](ancestors = 167155269)